src/trie.js

```diff
@@ -1,102 +1,518 @@
+  const visit = (R) => {
+    if (done.has(R)) {
+      return; // Skip, if already in the ordering
+    }
+    // If R is on the stack, we have a cycle...
+    if (seen.has(R)) {
+      const stack = [...seen]; // entries in Set are ordered by insertion order (neat)
+      const cycle = [...stack.slice(stack.indexOf(R)), R].map(r => r.pattern);
+      // Throw an error with the cycle
+      const err = new Error(`Roles may not contain dependency cycles, found: '${cycle.join('\' -> \'')}'`);
+      err.code = 'DependencyCycleError';
+      err.cycle = cycle;
+      throw err;
     }
+    // Mark R as seen, and visit it's dependencies
+    seen.add(R);
+    for (const dep of dependencies(R)) {
+      visit(dep);
+    }
+    // Add R to the ordering, and mark it as done
+    ordering.push(R);
+    done.add(R);
+  };
+
+  // Ensure that all rules are visited
+  for (const R of rules) {
+    visit(R);
+  }
+
+  // return our sorted value
+  return ordering;
+};
+
+// Export dependencyOrdering
+exports.dependencyOrdering = dependencyOrdering;
+
+/**
+ * Construct a trie where the given prefix have been pre-consumed.
+ *
+ * That is to say return a transformed trie, such that:
+ *   execute(withPrefix(node, prefix), input)
+ * is equivalent to:
+ *   execute(node, prefix + input)
+ *
+ * This method doesn't not interpret '*' at the end of prefix as kleene. Hence,
+ * the above equivalence does not hold for prefix = '...*' and input = ''.
+ * In practice we forbid rules on the form 'A*<..>' because '*' is ambiguous,
+ * hence, in withPrefix(node, 'A*') we shall never interpret '*' as kleene.
+ *
+ * Notice, this will NOT modify the input trie, but some of its children may be
+ * referenced in the result of this method. Hence, modifying the resulting trie
+ * will invalidate the input trie, and vice versus.
+ */
+const withPrefix = (trie, prefix = '') => {
+  const enter = new ScopeSetBuilder({optionallyClone: true});
+  const paramed = new ScopeSetBuilder({optionallyClone: true});
 
-    // a terminal star in the input here (state['*']['end']) will match all
-    // patterns, but we have already matched the `*` patterns (case A), so we
-    // can skip those.
-    while (i < n && rules[i].pattern[k] === '*' && rules[i].pattern.length === k + 1) {
-      i++;
+  // Find scopes granted along the path of the prefix
+  enter.add(trie.enter);
+  paramed.add(withParam(trie.paramed, prefix + PARAM));
+  // Traverse prefix from trie
+  const target = traverse(trie, prefix, (node, path) => {
+    // Prefix <..> with the remainder of the path
+    const remainder = prefix.slice(path.length);
+    enter.add(node.enter);
+    paramed.add(withParam(node.paramed, remainder + PARAM));
+  });
+
+  // if there is a target we have to add it's scopes too
+  let end = [];
+  let kleeneOnly = [];
+  if (target) {
+    end = target.end;
+    kleeneOnly = target.kleeneOnly;
+    enter.add(target.enter);
+    paramed.add(target.paramed); // <..> prefixed with empty remainder
```

Comment:
This comment baffled me until I worked it out myself.  How about `// equivalent to withParam(target.paramed, '' + PARAM)`